src/parser/parser.ts

```diff
@@ -115,7 +126,26 @@ export function parseAt(source: string, num: number) {
 export function parse(source: string, context: Context) {
   let program: es.Program | undefined
   try {
-    program = acornParse(source, createAcornParserOptions(context)) as unknown as es.Program
+    if (context.variant === Variant.TYPED) {
+      // For Source Typed variant, the code is parsed twice, first using the custom TypeParser
+      // (Acorn parser with plugin that allows for parsing of TS syntax), then using Babel Parser.
+      // This is a workaround as the custom TypeParser does not cover all type annotation cases needed for Source Typed
+      // and the Babel Parser does not allow for no semicolon/trailing comma errors when parsing.
+      program = TypeParser.parse(
+        source,
+        createAcornParserOptions(context)
+      ) as unknown as NodeWithTypeAnnotation<es.Program>
+      program = babelParse(source, {
+        sourceType: 'module',
+        plugins: ['typescript', 'estree']
+      }).program as unknown as NodeWithTypeAnnotation<es.Program>
```

Comment:
Maybe it would be good to add a comment here explaining why you cast the Babel types to estree types as we discussed offline?